src/pulsar_producer.erl

```diff
@@ -279,3 +431,185 @@ next_sequence_id(State = #state{sequence_id = SequenceId}) ->
     State#state{sequence_id = SequenceId+1}.
 
 log_error(Fmt, Args) -> logger:error("[pulsar-producer] " ++ Fmt, Args).
+
+invoke_callback(Callback, Resp) ->
+    invoke_callback(Callback, Resp, _BatchLen = 1).
+
+invoke_callback(_Callback = undefined, _Resp, _BatchLen) ->
+    ok;
+invoke_callback({M, F, A}, Resp, BatchLen) ->
+    lists:foreach(
+      fun(_) ->
+        erlang:apply(M, F, [Resp] ++ A)
+      end,  lists:seq(1, BatchLen));
+invoke_callback(Callback, Resp, BatchLen) when is_function(Callback, 1) ->
+    lists:foreach(
+      fun(_) ->
+        Callback(Resp)
+      end,  lists:seq(1, BatchLen)).
+
+queue_item_sizer(?Q_ITEM(_CallId, _Ts, _Batch) = Item) ->
+    size(queue_item_marshaller(Item)).
```

Comment:
we need a better size estimation.
serialising it only to get size is a waste of resource.
maybe: https://www.erlang.org/doc/man/erlang.html#external_size-1

Reply:
:heavy_check_mark: